### PR TITLE
[rush-lib] Fix overrides settings and package extensions conditions

### PR DESCRIPTION
**Pull request by pedro-gomes-92**
## Summary

Consider subspace pnpm `globalOverrides` and `globalPackageExtensions` during install & update

### What happened?

During rush installation `rush install -t .`, some subspace lock files were being considered "out of date" with the warning: "The overrides settings doesn't match the current shrinkwrap.", even though nothing was changed.

We identified that one of the subspaces had ~10 global overrides (e.g. common/config/subspaces/unified_wallet/pnpm-config.json), which were the **same** as stated in its shrinkwrap file (e.g. common/config/subspaces/unified_wallet/pnpm-lock.yaml). BUT the warning was still being logged for this subspace.

### What I expected to happen?

Rush should consider each subspace pnpm-config.json file during shrinkwrap up-to-date evaluation. 
I would expect no warning being logged, if the subspace shrinkwrap `overrides` parameter is synched with the pnpm-config `globalOverrides` parameter.

### The solution

Update overrides settings condition in `WorkspaceInstallManager`, by choosing between the objects:
- Repository pnpm-config `globalOverrides`
- Current subspace pnpm-config `globalOverrides`

Update package extension hash condition in `WorkspaceInstallManager`, by choosing between the objects:
- Repository pnpm-config `globalPackageExtensions`
- Current subspace pnpm-config `globalPackageExtensions`

## How it was tested

1. Create a repository with subspaces enabled: https://rushjs.io/pages/advanced/subspaces/
2. Update common/config/subspaces/default/pnpm-config.json, by adding a new entry into `globalOverrides`
3. Run `rush update --subspace default`

The shrinkfile `overrides` is now updated with the new added `globalOverrides` entry.

4. Run `rush install --subspace default`

The warning: "The overrides settings doesn't match the current shrinkwrap." will be shown in the terminal. This warning is incorrect.

## Impacted documentation

https://rushjs.io/pages/advanced/subspaces/

**Related issue:** https://github.com/microsoft/rushstack-websites/pull/235

### CODE REVIEW

File: libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts [hunk 1]
```diff
@@ -372,8 +373,16 @@ export class WorkspaceInstallManager extends BaseInstallManager {
     }
 
     // Check if overrides and globalOverrides are the same
+    const repoPnpmConfiguration: PnpmOptionsConfiguration = this.rushConfiguration.pnpmOptions;
+    const subspacePnpmConfiguration: PnpmOptionsConfiguration | undefined = subspace.getPnpmOptions();
+
+    const mergedGlobalOverrides: Record<string, string> = {
+      ...(repoPnpmConfiguration.globalOverrides ?? {}),
+      ...(subspacePnpmConfiguration?.globalOverrides ?? {})
+    };
```
Comment (iclanton) [inline libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts:382]:
```suggestion
    const mergedGlobalOverrides: Record<string, string> = {
      ...repoPnpmConfiguration.globalOverrides,
      ...subspacePnpmConfiguration?.globalOverrides
    };
```

The spread operator when assigning to an object handles `undefined`.

**Comment by pedro-gomes-92** [comment]
> Is this the only place where something like this is relevant, and is `globalOverrides` the only option on `pnpmOptions` that needs to be merged like this?

No, I just realised that `globalPackageExtensions` also need to be handled. Also it seems we should not merge config global configurations with subspace configurations, but they should be treated independently. Check `libraries/rush-lib/src/logic/installManager/InstallHelpers.ts:53`

```
const pnpmOptions: PnpmOptionsConfiguration =
  subspace.getPnpmOptions() || rushConfiguration.pnpmOptions;
if (!commonPackageJson.pnpm) {
  commonPackageJson.pnpm = {};
}

if (pnpmOptions.globalOverrides) {
  commonPackageJson.pnpm.overrides = pnpmOptions.globalOverrides;
}
```


> And would it make more sense for the subspace pnpm options to be pre-merged with the globals?

Since they should be handled independently, I don't think so.

**Comment by iclanton** [comment]
>Also it seems we should not merge config global configurations with subspace configurations, but they should be treated independently.

Hmm. That's not how I'd expect these config files to work. The code in `InstallHelpers` seems wrong. It'd be good to get some opinions from @william2958, who introduced that code.

**Comment by octogonz** [comment]
> > Also it seems we should not merge config global configurations with subspace configurations, but they should be treated independently.
> 
> Hmm. That's not how I'd expect these config files to work. The code in `InstallHelpers` seems wrong. It'd be good to get some opinions from @william2958, who introduced that code.

The official specification should be found on the Rush website docs.  Seems like I had created this PR a while ago, but it never got merged:

https://github.com/microsoft/rushstack-websites/pull/235

According to that doc, the settings do NOT get merged:

> - Subspace config folder: `common/config/subspaces/<subspace-name>/`
> - Monorepo config folder: `common/config/rush/`
>
>
> | Subspace file                                                    | Monorepo file                                                | Inheritance                                                                                                                                                                                      |
> | :--------------------------------------------------------------- | :----------------------------------------------------------- | :----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
> | `pnpm-config.json`                                               | `pnpm-config.json` | **Fallback**: The monorepo file is used only if the subspace's file is absent.                                                                                                                   |

Let's double-check the design and make sure the behavior is clearly documented before merging PR #5015.

**Comment by octogonz** [comment]
> Hmm. That's not how I'd expect these config files to work. The code in `InstallHelpers` seems wrong.

@iclanton The `pnpm-config.json` file contains:

- settings like `globalPackageExtensions` represent a huge pile of hacks and adjustments to fix version conflicts in the PNPM lockfile
- settings like `strictPeerDependencies` and `useWorkspaces` that are crucial information for anyone who is investigating a lockfile issue

I think this is why we were reluctant for a design that requires engineers to look in two different files (and mentally merge those results) to understand the final configuration.  Whereas a fallback is more reasonable -- if the file is completely absent, then you can look in the other place, and that file alone is the whole story of your configuration.

**Comment by pedro-gomes-92** [comment]
> Let's double-check the design and make sure the behavior is clearly documented before merging PR #5015.

Regarding the design approach, I am a bit mixed to choose between MERGE vs SELECT. 

As a first impression, I would expect the act like a MERGE, because of how other tools "normally" do it (e.g. eslint, tsconfig, webpack). It is true that these tools still need to explicitly add the parameter `extends` to take advantage of the inheritance approach. I would propose a similar design feature for our subspace configurations:

```javascript
// config/subspaces/s1/pnpm-config.json
{
  "extends": "../../rush/pnpm-config.json",
  "globalOverrides": {
    "testA": "1.0.0"
  }
}
```

This will allow developers to fully understand how the configuration inheritance works without further guessing the expected results (like I initially did). At the same time, subspace configuration files can be simplified, by removing duplicated configuration values, improving readability and consistency.

**Comment by octogonz** [comment]
> I would expect to act like a MERGE, because of how other tools "normally" do it (e.g. eslint, tsconfig, webpack).

We discussed this question, and for now are going to keep the current behavior. Key points:

1. **We need a motivating example.**  The original motivation of subspaces was to enable multiple lockfiles to be managed independently, often by teams who actually wish to minimize their interaction. In other words, those teams prefer duplicated work versus having to coordinate their version conflict management.  Thus, if we provide an merging model (beyond the simple fallback described above), it would be helpful to have a real world example showing why copy+paste has become cumbersome.

2. **Inheritance has multiple possible designs.**  Specifically, it's not required that we need to model this as merging `pnpm-config.json` files.  Certain settings such as `useWorkspaces` are confusing to merge -- it would be better for them to have a single home. Other settings such as  `pnpmLockfilePolicies.disallowInsecureSha1.exemptPackageVersions` might require complex merging rules.  These complexities could be reduced by inventing a new file format (e.g. `pnpm-shared-config.json`) that only includes the settings that really make sense for inheritance.   But in order to design that, we need a motivating example.

3. **Naive `extends` raises performance concerns.** Every time Rush is invoked (including TAB completion), it needs to load quite a lot of config files.  We've worked hard to optimize this.  As a result, Rush has avoided the usual `"extends"` mechanism, as it enables developers to arbitrarily increase the number of files that must be read.  (On the other hand, tools like Heft can rely heavily on `"extends"` because they are invoked per-project, thus their overhead is parallelized somewhat.)  Thus if we add support for inheritance, it might be better to choose a design that prevents endless chaining as is possible with `"extends"`.

**Comment by octogonz** [comment]
> The official specification should be found on the Rush website docs. Seems like I had created this PR a while ago, but it never got merged:

https://github.com/microsoft/rushstack-websites/pull/235 has been merged

**Review by octogonz — APPROVED** [review]

**Pull request merged**